xds/src/main/java/io/grpc/xds/client/ControlPlaneClient.java

```diff
@@ -232,27 +249,50 @@ void readyHandler() {
   private void startRpcStream() {
     checkState(adsStream == null, "Previous adsStream has not been cleared yet");
     adsStream = new AdsStream();
+    adsStream.start();
     logger.log(XdsLogLevel.INFO, "ADS stream started");
     stopwatch.reset().start();
   }
 
+  void sendDiscoveryRequests() {
+    if (adsStream == null) {
+      startRpcStream();
+      // when the stream becomes ready, it will send the discovery requests
+      return;
+    }
+
+    if (isConnected()) {
+      adjustAllResourceSubscriptions();
+    }
+  }
+
+  void adjustAllResourceSubscriptions() {
+    if (isInBackoff()) {
+      return;
+    }
+
+    Set<XdsResourceType<?>> subscribedResourceTypes =
+        new HashSet<>(resourceStore.getSubscribedResourceTypesWithTypeUrl().values());
+
+    for (XdsResourceType<?> type : subscribedResourceTypes) {
+      adjustResourceSubscription(type);
+    }
+  }
+
   @VisibleForTesting
   public final class RpcRetryTask implements Runnable {
     @Override
     public void run() {
-      if (shutdown) {
+      logger.log(XdsLogLevel.DEBUG, "Retry timeout. Restart ADS stream {0}", logId);
+      if (shutdown || isReady()) {
```

Comment:
How can `isReady() == true` here be significant? The only thing this does is create `adsStream`, and `isReady() == false` when `adsStream == null`. I think the `isReady()` here should be deleted, but also you now need to move cancelling `rpcRetryTimer` from `readyHandler()` to `startRpcStream()`. Then this will never be called when `adsStream == null` and you can remove the condition around `startRpcStream()` that you added in this PR.

Reply:
Done